app/controllers/admin/flags_controller.rb

```diff
@@ -71,65 +62,70 @@ def agree
       user: current_user
     )
 
-    post_action_type = PostAction.post_action_type_for_post(post.id)
-
-    if !post_action_type
-      render_json_error(
-        I18n.t("flags.errors.already_handled"),
-        status: 409
-      )
-      return
-    end
+    reviewable = post.reviewable_flag
+    return render_json_error(I18n.t("flags.errors.already_handled"), status: 409) if reviewable.blank?
 
     keep_post = ['silenced', 'suspended', 'keep'].include?(params[:action_on_post])
     delete_post = params[:action_on_post] == "delete"
     restore_post = params[:action_on_post] == "restore"
 
     if delete_post
-      # PostDestroy calls PostAction.agree_flags!
+      # PostDestroy automatically agrees with flags
       destroy_post(post)
     elsif restore_post
-      PostAction.agree_flags!(post, current_user, delete_post)
-      PostDestroyer.new(current_user, post).recover
+      reviewable.perform(current_user, :agree_and_restore)
     else
-      PostAction.agree_flags!(post, current_user, delete_post)
-      if !keep_post
-        PostAction.hide_post!(post, post_action_type)
-      end
+      reviewable.perform(
+        current_user,
+        :agree_and_keep,
+        post_was_deleted: delete_post,
+        hide_post: !keep_post
+      )
     end
 
     render body: nil
   end
 
   def disagree
+    Discourse.deprecate("FlagsController#disagree has been deprecated, please use the Reviewable API instead")
     params.permit(:id)
     post = Post.find(params[:id])
 
-    DiscourseEvent.trigger(
-      :before_staff_flag_action,
-      type: 'disagree',
-      post: post,
-      user: current_user
-    )
+    if reviewable = post&.reviewable_flag
+      DiscourseEvent.trigger(
+        :before_staff_flag_action,
+        type: 'disagree',
+        post: post,
+        user: current_user
+      )
 
-    PostAction.clear_flags!(post, current_user)
+      if post.hidden?
+        reviewable.perform(current_user, :disagree_and_restore)
+      else
+        reviewable.perform(current_user, :disagree)
+      end
+    end
 
     render body: nil
   end
 
   def defer
+    Discourse.deprecate("FlagsController#defer has been deprecated, please use the Reviewable API instead")
+
     params.permit(:id, :delete_post)
     post = Post.find(params[:id])
 
-    DiscourseEvent.trigger(
-      :before_staff_flag_action,
-      type: 'defer',
-      post: post,
-      user: current_user
-    )
+    if reviewable = post&.reviewable_flag
+      DiscourseEvent.trigger(
+        :before_staff_flag_action,
+        type: 'defer',
+        post: post,
+        user: current_user
+      )
```

Comment:
That's code I didn't change from before. I do agree it could be put into a method but in this case the entire controller is deprecated and going to be deleted anyway, so probably not worth it.